subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheGradlePropertiesIntegrationTest.groovy

```diff
@@ -179,15 +179,52 @@ class ConfigurationCacheGradlePropertiesIntegrationTest extends AbstractConfigur
         configurationCache.assertStateStored()
 
         when:
-        System.clearProperty('fromPropertiesFile')
+        System.clearProperty("$systemProp")
         configurationCacheRun fixture.task
 
         then:
-        outputDoesNotContain "because system property 'fromPropertiesFile' has changed"
+        outputDoesNotContain "because system property '$systemProp' has changed"
         outputContains fixture.expectedOutputBeforeChange
         configurationCache.assertStateLoaded()
 
         cleanup:
-        System.clearProperty('fromPropertiesFile')
+        System.clearProperty("$systemProp")
+    }
+
+    def "same system property from multiple included build properties files"() {
```

Comment:
I'd add the test checking the command-line override for the system property, as it wins over the `gradle.properties` value. Adding the override should invalidate the configuration cache - but if there are property reads at the configuration time. I'm not sure if the current implementation of the `EvironmentChangeTracker` can handle this correctly, though - some non-trivial amendments may be necessary.

There is a TODO in DefaultGradlePropertiesLoader to test if `included-build/gradle.properties` overrides `gradle.properties` of the root build; it might be a good time and place to add an actual test for this.

One more tricky scenario: what if we read the system property inside the `settings.gradle`? In the root build or in one of the included builds? It affects the output, but shouldn't affect the value used at execution time.

Reply:
> I'd add the test checking the command-line override for the system property, as it wins over the `gradle.properties` value. Adding the override should invalidate the configuration cache - but if there are property reads at the configuration time. I'm not sure if the current implementation of the `EvironmentChangeTracker` can handle this correctly, though - some non-trivial amendments may be necessary.

But in this case startParameters of build is changed, shouldn't we drop an entry regardless(as we doing with gradleProperties) was systemProp used in config time?

Reply:
Unlike Gradle properties, tracking of system properties is precise, so if no reads of said property happened at configuration time, its value doesn't affect the work graph. We should still respect the value when executing the build itself, I can see how restoring the system property in the EnvironmentChangeTracker can get in the way, overriding the command-line value with the cached one.